internal/storage/bucket_handle_test.go

```diff
@@ -168,6 +168,44 @@ func (t *BucketHandleTest) TestNewReaderMethodWithInvalidGeneration() {
 	AssertEq(nil, rc)
 }
 
+func (t *BucketHandleTest) TestNewReaderMethodWithCompressionEnabled() {
+	rc, err := t.bucketHandle.NewReader(context.Background(),
+		&gcs.ReadObjectRequest{
+			Name: TestGzipObjectName,
+			Range: &gcs.ByteRange{
+				Start: uint64(0),
+				Limit: uint64(len(ContentInTestGzipObject)),
+			},
+			ReadCompressed: true,
+		})
+
+	AssertEq(nil, err)
+	defer rc.Close()
+	buf := make([]byte, len(ContentInTestGzipObject))
+	_, err = rc.Read(buf)
+	AssertEq(nil, err)
+	ExpectEq(ContentInTestGzipObject, string(buf[:]))
+}
+
+func (t *BucketHandleTest) TestNewReaderMethodWithCompressionDisabled() {
+	rc, err := t.bucketHandle.NewReader(context.Background(),
+		&gcs.ReadObjectRequest{
+			Name: TestGzipObjectName,
+			Range: &gcs.ByteRange{
+				Start: uint64(0),
+				Limit: uint64(len(ContentInTestGzipObject)),
+			},
+			ReadCompressed: false,
```

Comment:
shouldn't the read fail here?

Reply:
@vadlakondaswetha 

Yes, it should have failed. This test case basically represents the scenario before the change. 

It's probably something to do with either the setup of fakeStorage (meaning somehow its read/write operations don't mimic those of GCS or error isn't being propagated up the stack) or that of the TestGzipObject (meaning somehow the content-encoding isn't being propagated to the actual fakeBucket) . I'm unable to figure out at the moment.

Is this test a deal-breaker? If not, I'd like to omit it now for the release and investigate it later.

Reply:
Look at the fake storage implemetation and see what is happening.

Reply:
Add a test zip file and add unit test case reading it.

Reply:
I tested by adding a gzip file (download of a file uploaded by gsutil cp -Z in the unit test and the behavior didn't change. storage client still didn't decompress the file.
I ran the same function in a standalone go app using the same set up (same go client version too). In this standalone, the object was decompressed correctly.

I debugged into the place where the request and responses are captured in the httpclient for both programs parallely to see any difference. The requests sent are same for both except for the bucket name and the object path.

I diffed the responses and found two major differences, which might give us a hint.

In the standalone, response proto major is http:2.0, which brings with a lot of new headers X-goog-*, which I remember Frank told me were added to handle gzip behavior. By contrast, response proto in our unit test is http:11, which doesn't bring most of the above X-goog headers,
response has a header/attribute called Uncompressed. In the standalone, this attribute is true, but in our unit test, it is false.

@vadlakondaswetha  This is all I have for now. Any suggestions ?

Reply:
Did you check the fake storage server code what it is doing. is it handling gzip files correctly or not. Unit tests is not calling gcs but calling fake storage server.

Reply:
@vadlakondaswetha thanks. That helped.

So, fake storage server is in https://github.com/fsouza/fake-gcs-server . The specific function/API invoked on trying to read an object comes [here](https://github.com/fsouza/fake-gcs-server/blob/3292d72191cbf38d472095d28461c5986739e5cf/fakestorage/object.go#L795) . The code for decompressing content-encoding gzip data was implemented in release version [1.40.3](https://github.com/fsouza/fake-gcs-server/releases/tag/v1.40.3) of this library and our current dependency is linked to version [1.38.4](https://github.com/GoogleCloudPlatform/gcsfuse/blob/d4e07940eb1d5dac80d12a200bea8b04838ab7f3/go.mod#L10C6-L10C6) . TBC.

Reply:
I tested that by switching the version of github.com/fsouza/fake-gcs-server to 1.40.3 makes the test behave as it should. Though I haven't tested the impact of that on the other tests, functionalities etc.

We can do one of the following:
1. Update the version of fsouza/fake-gcs-server in our go.mod to 1.40.3 or may be even 1.47.4(the latest right now) and test that the dependency change doesn't break anything else.
2. Disable the test now, and enable it with the change in (1) above after the release.

@vadlakondaswetha let me know what your call is.

Reply:
Captured in https://github.com/GoogleCloudPlatform/gcsfuse/pull/1255#issuecomment-1674484952 .